### ISSUE TITLE
Azure Metrics Output fails with negative values

### ISSUE THREAD

**Issue by dstengele** [influxdata/telegraf#10865]
### Relevant telegraf.conf

```toml
[global_tags]
[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  flush_jitter = "0s"
  precision = ""
  hostname = "vm-jira-prd-01"
  omit_hostname = false
[[outputs.azure_monitor]]
[[inputs.jolokia2_agent]]
  urls = ["http://localhost:8085/jira/plugins/servlet/jolokia"]
  [[inputs.jolokia2_agent.metric]]
    name  = "jvm_memory"
    mbean = "java.lang:type=Memory"
    paths = ["HeapMemoryUsage", "NonHeapMemoryUsage", "ObjectPendingFinalizationCount"]
```

### Logs from Telegraf

```
2022-03-22T11:55:06Z I! Starting Telegraf 1.21.4
2022-03-22T11:55:06Z I! Loaded inputs: jolokia2_agent
2022-03-22T11:55:06Z I! Loaded aggregators:
2022-03-22T11:55:06Z I! Loaded processors:
2022-03-22T11:55:06Z I! Loaded outputs: azure_monitor
2022-03-22T11:55:06Z I! Tags enabled: host=vm-jira-prd-01
2022-03-22T11:55:06Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"vm-jira-prd-01", Flush Interval:10s
2022-03-22T11:55:06Z D! [agent] Initializing plugins
2022-03-22T11:55:06Z D! [agent] Connecting outputs
2022-03-22T11:55:06Z D! [agent] Attempting connection to [outputs.azure_monitor]
2022-03-22T11:55:06Z D! [outputs.azure_monitor] Writing to Azure Monitor URL: https://westeurope.monitoring.azure.com/subscriptions/$UUID/resourceGroups/$RESOURCE_GROUP/providers/Microsoft.Compute/virtualMachines/vm-jira-prd-01/metrics
2022-03-22T11:55:06Z D! [agent] Successfully connected to outputs.azure_monitor
2022-03-22T11:55:06Z D! [agent] Starting service inputs
2022-03-22T11:55:16Z D! [outputs.azure_monitor] Buffer fullness: 0 / 10000 metrics
2022-03-22T11:55:26Z D! [outputs.azure_monitor] Buffer fullness: 0 / 10000 metrics
2022-03-22T11:55:36Z D! [outputs.azure_monitor] Buffer fullness: 0 / 10000 metrics
2022-03-22T11:55:46Z D! [outputs.azure_monitor] Buffer fullness: 0 / 10000 metrics
2022-03-22T11:55:56Z D! [outputs.azure_monitor] Buffer fullness: 0 / 10000 metrics
2022-03-22T11:56:06Z D! [outputs.azure_monitor] Buffer fullness: 9 / 10000 metrics
2022-03-22T11:56:06Z E! [agent] Error writing to outputs.azure_monitor: failed to write batch: [400] 400 Bad Request
2022-03-22T11:56:16Z D! [outputs.azure_monitor] Buffer fullness: 9 / 10000 metrics
2022-03-22T11:56:16Z E! [agent] Error writing to outputs.azure_monitor: failed to write batch: [400] 400 Bad Request
2022-03-22T11:56:26Z D! [outputs.azure_monitor] Buffer fullness: 9 / 10000 metrics
2022-03-22T11:56:26Z E! [agent] Error writing to outputs.azure_monitor: failed to write batch: [400] 400 Bad Request
^C2022-03-22T11:56:29Z D! [agent] Stopping service inputs
2022-03-22T11:56:29Z D! [agent] Input channel closed
2022-03-22T11:56:29Z I! [agent] Hang on, flushing any cached metrics before shutdown
2022-03-22T11:56:29Z D! [outputs.azure_monitor] Buffer fullness: 9 / 10000 metrics
2022-03-22T11:56:29Z E! [agent] Error writing to outputs.azure_monitor: failed to write batch: [400] 400 Bad Request
2022-03-22T11:56:29Z I! [agent] Stopping running outputs
2022-03-22T11:56:29Z D! [agent] Stopped Successfully
```

### System info

Telegraf 1.21.4, Ubuntu 20.04 LTS in Azure, outputs.azure_monitor configured via Managed Service Identity

### Steps to reproduce

1. Read JVM Memory Usage (e.g. via inputs.jolokia2_agent)
2. Have one or more of the values undefined (-1, see [Oracle-Docs](https://docs.oracle.com/javase/8/docs/api/java/lang/management/MemoryUsage.html#MemoryUsage-long-long-long-long-))
3. Submit the metrics to Azure Metrics via outputs.azure_monitor

### Expected behavior

Negative Values can be submitted correctly to Azure Metrics. No error is thrown.

### Actual behavior

Azure API returns 400 Bad Request with the following body:

```json
{
  "error": {
    "code": "BadRequest",
    "message": "'series' cannot have 'sum' less than 'max'\r\n"
  }
}
```

### Additional info

I had to patch the azure_monitor output plugin to log the response body, see #8916.

**Comment by powersj** [comment]
Hi,

> I had to patch the azure_monitor output plugin to log the response body, see https://github.com/influxdata/telegraf/issues/8916.

I have put up #10866 to include the error message body when issues arise. Feel free to propose a PR with a patch should you run into something like this again.

> Azure API returns 400 Bad Request with the following body:

Off-hand, I am not sure what Telegraf itself can do about this. Did you have a suggestion? Have you considered using a processor to modify your own data?

Thanks!

**Comment by dstengele** [comment]
Hi,

thank you for the hint to processors, I was not aware that those existed. :D I added the following processor config:
```
[[processors.execd]]
  command = ["sed", "s/=-[0-9]*i/=0i/"]
```
It's not really pretty, but at least for our use case it works.

Regarding the PR, I do not usually use Go, so I was not feeling confident enough for a PR. I will try next time though.

I would suggest that the azure_monitor output plugin should check for negative values and warn the user as otherwise the error is rather opaque, even when printing the response body. I had to look through the request body to find the series and why sum was less than max.

I will try to create a PR for that as soon as I find the time.

**Comment by powersj** [comment]
> thank you for the hint to processors, I was not aware that those existed. :D I added the following processor config:

Awesome

> It's not really pretty, but at least for our use case it works.

Understood, the processors are great for handling these types of things where making a global change to Telegraf or to a specific dataset doesn't make sense.

> I would suggest that the azure_monitor output plugin should check for negative values and warn the user as otherwise the error is rather opaque, even when printing the response body. I had to look through the request body to find the series and why sum was less than max.

Do you think having my fix in #10866 would alleviate the work you had to put in to figure out what was wrong?

Thanks again!

**Comment by phoehnel** [comment]
Came across this via google.

The sed from @dstengele is great, but i sometimes had negative values that where not int and thus not suffixed with i so that the sed mentioned here will not work, e.g.:
```
xxx,host=vm-xxx-prd,wifi-tx-bytes_rate=-1592.3333333333333 [...]
xxx,host=vm-xxx-prd,wifi-signal=-78.1f [...]
```

**This version worked for all values i had:**
```
# needed because azure monitor does not support negative values
[[processors.execd]]
  command = ["sed", "s/=-[0-9.]*\\([fi]*\\)/=0\\1/g"]
```

If you want to **keep the absolute value** of certain measurements instead of setting them to 0 you can use:

```
# needed because azure monitor does not support negative values
[[processors.execd]]
  command = ["sed","s/=-\\([0-9.]*\\)\\([fi]*\\)/=\\1\\2/g"]

  fieldinclude = ["wifi-signal*", "*x-bytes_rate"]
```
